spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/core/SpannerTemplate.java

```diff
@@ -207,68 +208,87 @@ private StringBuilder wrapAsSubSelect(String sql) {
 
 	@Override
 	public void insert(Object object) {
-		applyMutationsUsingEntity(this.mutationFactory::insert, object);
+		applyMutations(this.mutationFactory.insert(object));
+	}
+
+	@Override
+	public void insertAll(Iterable objects) {
+		applyMutations(
+				getListCollectionFunction(this.mutationFactory::insert).apply(objects));
 	}
 
 	@Override
 	public void update(Object object) {
-		applyMutationsTwoArgs(this.mutationFactory::update, object, null);
+		applyMutations(this.mutationFactory.update(object, null));
+	}
+
+	@Override
+	public void updateAll(Iterable objects) {
+		applyMutations(
+				getListCollectionFunction(x -> this.mutationFactory.update(x, null))
+						.apply(objects));
 	}
 
 	@Override
 	public void update(Object object, String... includeColumns) {
-		applyMutationsTwoArgs(this.mutationFactory::update, object,
+		applyMutations(
+				this.mutationFactory.update(object,
 				includeColumns.length == 0 ? null
-						: Optional.of(new HashSet<>(Arrays.asList(includeColumns))));
+						: Optional.of(new HashSet<>(Arrays.asList(includeColumns)))));
 	}
 
 	@Override
 	public void update(Object object, Optional<Set<String>> includeColumns) {
-		applyMutationsTwoArgs(this.mutationFactory::update, object, includeColumns);
+		applyMutations(this.mutationFactory.update(object, includeColumns));
 	}
 
 	@Override
 	public void upsert(Object object) {
-		applyMutationsTwoArgs(this.mutationFactory::upsert, object, null);
+		applyMutations(this.mutationFactory.upsert(object, null));
+	}
+
+	@Override
+	public void upsertAll(Iterable objects) {
+		applyMutations(
+				getListCollectionFunction(x -> this.mutationFactory.upsert(x, null))
+						.apply(objects));
```

Comment:
You always immediately call apply(objects) on the result of getListCollectionFunction. I think it would be better to pass  the objects to getListCollectionFunction along with individualEntityMutationFunc and return a list of mutations back instead of returning a function.

Reply:
change made, PTAL